### ISSUE TITLE
MSB3277 warnings when referencing Microsoft.AspNetCore.Server.Kestrel.Core and System.Security.Cryptography.Cng

### ISSUE THREAD

**Issue by egraff** [dotnet/aspnetcore#36619]
I'm opening a new issue for https://github.com/dotnet/aspnetcore/issues/34299, because the former issue was closed because I did not reply quick enough while I had my summer vacation, and after I did reply, I got no answer and the issue was locked.

---

### Description

When building a _multi-targeted_ project that includes netcoreapp2.1 as one of the target frameworks, I get MSB3277 warnings if referencing a version > 4.5.2 of System.Security.Cryptography.Cng when including version 2.1.25 of Microsoft.AspNetCore.Server.Kestrel.Core.

If I downgrade to version 2.1.7 of Microsoft.AspNetCore.Server.Kestrel.Core, I can use newer versions of System.Security.Cryptography.Cng (such as 4.7.0 or 5.0.0) without warnings.

The warnings look like this:
```
C:\Program Files\dotnet\sdk\5.0.202\Microsoft.Common.CurrentVersion.targets(2182,5): warning MSB3277: Found conflicts between different versions of "System.Security.Cryptography.Cng" that could not be resolved. [C:\path\to\problem.csproj]
C:\Program Files\dotnet\sdk\5.0.202\Microsoft.Common.CurrentVersion.targets(2182,5): warning MSB3277: There was a conflict between "System.Security.Cryptography.Cng, Version=4.3.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" and "System.Security.Cryptography.Cng, Version=4.3.1.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a". [C:\path\to\problem.csproj]
C:\Program Files\dotnet\sdk\5.0.202\Microsoft.Common.CurrentVersion.targets(2182,5): warning MSB3277:     "System.Security.Cryptography.Cng, Version=4.3.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" was chosen because it was primary and "System.Security.Cryptography.Cng, Version=4.3.1.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" was not. [C:\path\to\problem.csproj]
C:\Program Files\dotnet\sdk\5.0.202\Microsoft.Common.CurrentVersion.targets(2182,5): warning MSB3277:     References which depend on "System.Security.Cryptography.Cng, Version=4.3.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" [C:\Users\Erlend\.nuget\packages\system.security.cryptography.cng\4.6.0\ref\netcoreapp2.1\System.Security.Cryptography.Cng.dll]. [C:\path\to\problem.csproj]
C:\Program Files\dotnet\sdk\5.0.202\Microsoft.Common.CurrentVersion.targets(2182,5): warning MSB3277:         C:\Users\Erlend\.nuget\packages\system.security.cryptography.cng\4.6.0\ref\netcoreapp2.1\System.Security.Cryptography.Cng.dll [C:\path\to\problem.csproj]
C:\Program Files\dotnet\sdk\5.0.202\Microsoft.Common.CurrentVersion.targets(2182,5): warning MSB3277:           Project file item includes which caused reference "C:\Users\Erlend\.nuget\packages\system.security.cryptography.cng\4.6.0\ref\netcoreapp2.1\System.Security.Cryptography.Cng.dll". [C:\path\to\problem.csproj]
C:\Program Files\dotnet\sdk\5.0.202\Microsoft.Common.CurrentVersion.targets(2182,5): warning MSB3277:             C:\Users\Erlend\.nuget\packages\system.security.cryptography.cng\4.6.0\ref\netcoreapp2.1\System.Security.Cryptography.Cng.dll [C:\path\to\problem.csproj]
C:\Program Files\dotnet\sdk\5.0.202\Microsoft.Common.CurrentVersion.targets(2182,5): warning MSB3277:     References which depend on "System.Security.Cryptography.Cng, Version=4.3.1.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" []. [C:\path\to\problem.csproj]
C:\Program Files\dotnet\sdk\5.0.202\Microsoft.Common.CurrentVersion.targets(2182,5): warning MSB3277:         C:\Users\Erlend\.nuget\packages\microsoft.aspnetcore.server.kestrel.core\2.1.25\lib\netcoreapp2.1\Microsoft.AspNetCore.Server.Kestrel.Core.dll [C:\path\to\problem.csproj]
C:\Program Files\dotnet\sdk\5.0.202\Microsoft.Common.CurrentVersion.targets(2182,5): warning MSB3277:           Project file item includes which caused reference "C:\Users\Erlend\.nuget\packages\microsoft.aspnetcore.server.kestrel.core\2.1.25\lib\netcoreapp2.1\Microsoft.AspNetCore.Server.Kestrel.Core.dll". [C:\path\to\problem.csproj]
C:\Program Files\dotnet\sdk\5.0.202\Microsoft.Common.CurrentVersion.targets(2182,5): warning MSB3277:             C:\Users\Erlend\.nuget\packages\microsoft.aspnetcore.server.kestrel.core\2.1.25\lib\netcoreapp2.1\Microsoft.AspNetCore.Server.Kestrel.Core.dll [C:\path\to\problem.csproj]
C:\Program Files\dotnet\sdk\5.0.202\Microsoft.Common.CurrentVersion.targets(2182,5): warning MSB3277:         C:\Users\Erlend\.nuget\packages\microsoft.netcore.app\2.1.0\ref\netcoreapp2.1\System.Core.dll [C:\path\to\problem.csproj]
C:\Program Files\dotnet\sdk\5.0.202\Microsoft.Common.CurrentVersion.targets(2182,5): warning MSB3277:           Project file item includes which caused reference "C:\Users\Erlend\.nuget\packages\microsoft.netcore.app\2.1.0\ref\netcoreapp2.1\System.Core.dll". [C:\path\to\problem.csproj]
C:\Program Files\dotnet\sdk\5.0.202\Microsoft.Common.CurrentVersion.targets(2182,5): warning MSB3277:             C:\Users\Erlend\.nuget\packages\microsoft.netcore.app\2.1.0\ref\netcoreapp2.1\System.Core.dll [C:\path\to\problem.csproj]
```

and the minimal csproj file needed to reproduce the problem looks like this:

```xml
<?xml version="1.0" encoding="utf-8"?>
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Library</OutputType>
    <TargetFrameworks>net48;netcoreapp2.1</TargetFrameworks>
  </PropertyGroup>
  <PropertyGroup>
    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
    <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.Server.Kestrel.Core" Version="2.1.25" />
    <PackageReference Include="System.Security.Cryptography.Cng" Version="4.7.0" />
  </ItemGroup>
</Project>
```

## Configuration
Output from dotnet --info:

```
.NET SDK (reflecting any global.json):
 Version:   5.0.202
 Commit:    db7cc87d51

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.202\

Host (useful for support):
  Version: 5.0.5
  Commit:  2f740adc14

.NET SDKs installed:
  5.0.202 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.27 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.27 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.27 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
 ```

---

In the previous issue, it was pointed out that I could [remove the binding redirects](https://github.com/dotnet/aspnetcore/issues/34299#issuecomment-878820334) to make the problem go away. However, I had not made it clear that it was a multi-targeted project, and I _do_ need the binding redirects for the .NET Framework target. Also, I would like to avoid having to do something like
```xml
<PropertyGroup Condition="'$(TargetFramework)' == 'net48'">
  <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
  <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
</PropertyGroup>
```

I have the binding redirect properties set in a lot of projects that are multi-targeted for multiple versions .NET Core and .NET Framework, and only in this particular case do I see warnings such as those described above.

**Comment by adityamandaleeka** [comment]
Triage: One possible explanation for why _downgrading_ Microsoft.AspNetCore.Server.Kestrel.Core makes the problem go away might be that the older Kestrel isn't (transitively) depending on a specific Cng.

That workaround (of having the binding redirects be conditionally set based on target framework) is likely the best option.

**Comment by egraff** [comment]
@adityamandaleeka I've done some more digging, and the problem is that Microsoft.AspNetCore.Server.Kestrel.Core 2.1.25 depends on assembly-version 4.3.1.1 of System.Security.Cryptography.Cng. As far as I can tell, this assembly-version does only exist in nuget-version 4.5.2 of System.Security.Cryptography.Cng. Now, version 4.5.2 was published as a dot-release on Jan 14 2020, so it's published "later" than e.g. version 4.7.0 which was released in 2019. But version 5.0.0, for example, was released in November 2020, but still has assembly-version 4.3.1.0.

The issue seems to be that System.Security.Cryptography.Cng got released with a version 4.3.1.1, when it probably should have stayed at version 4.3.1.0. By referencing nuget-version 4.5.2, thus referencing assembly-version 4.3.1.1, it becomes impossible to reference newer versions of System.Security.Cryptography.Cng without effectively *downgrading* the version.

Removing the binding redirects is not a solution - that simply hides the bug!

IMHO, the solution would be to ensure that the .NET Core 2.1 version of Microsoft.AspNetCore.Server.Kestrel.Core references assembly-version 4.3.1.0 of System.Security.Cryptography.Cng, regardless of which nuget-version is specified as dependency in the .nuspec.
